Support running Quarkus tests from Gradle

aloubyansky opened the pull request:
Fixes #2307, #3103 

This PR adds a Quarkus Gradle task that sets Quarkus-specific properties (such as `native.image.path` and newly added `quarkus-deployment-cp` whose value is a deployment classpath necessary to build the test) on the test tasks for them to be able to run.

Gradle projects will need to add the following to their build scripts
````
test {
    useJUnitPlatform()

    // @SubstrateTest and JVM mode tests can't be mixed in the same run
    forkEvery 1
}
````

aloubyansky committed:
Quarkus Gradle task that sets properties on the test tasks for them to be able to run. Among which the deployment classpath. This allows to run Quarkus tests from Gradle.

aloubyansky commented:
@garyttierney fyi, this is kind of a quick fix

aloubyansky commented:
@aalmiray do you mind if I ask you for a review? It's a small one but I would appreciate your opinion.
The idea behind this change is to resolve some test-related data/config before the tests get executed and pass that as system properties to the tests. Do you see any problem with this approach? Or perhaps suggestions to improve it? Thanks a lot!

geoand commented:
@aloubyansky I just tried the [getting-started](https://github.com/quarkusio/quarkus-quickstarts/blob/master/getting-started/build.gradle) quickstart with this PR, but I am seeing:

```
* What went wrong:
Execution failed for task ':compileTestJava'.
> Could not resolve all files for configuration ':testCompileClasspath'.
   > Could not find guice-no_aop.jar (com.google.inject:guice:4.1.0).
     Searched in the following locations:
         file:/home/gandrian/.m2/repository/com/google/inject/guice/4.1.0/guice-4.1.0-no_aop.jar
   > Could not find guice-no_aop.jar (com.google.inject:guice:4.1.0).
     Searched in the following locations:
         file:/home/gandrian/.m2/repository/com/google/inject/guice/4.1.0/guice-4.1.0-no_aop.jar

```

Any idea of this is related?

aloubyansky commented:
Ah, yes, Gradle enforces BOMs on tests as well, while Maven does not, unless scope test is explicitly configured. So, I  this case the Gradle build expects that version to be available locally but maven downloaded 4.2.0, or something.

geoand commented:
Thanks a lot for the explanation!

garyttierney commented:
> @garyttierney fyi, this is kind of a quick fix

I think that if we can resolve the issue above this might be the correct approach. Compared to this tree, my branch has a lot of code churn for the exact same effect -- pulling the AppModel dependencies out of the Gradle plugin remotely.

aalmiray commented:
@aloubiansky: sure, no problem, happy to help. I’ll add my comments by the end of this week at the latest

gsmet commented:
@aloubyansky @geoand I see we have an explanation for the failure but should I expect it to fail on the next release when this PR is in?

geoand commented:
@gsmet I personally don't know

Dufgui commented:
As I can see the missing dependency is particular, it have a classifier and used in the test scope.
I try to to remote debug the gradle task, but I fail. It compile in eclipse as a gradle project. I put gradle opts to remote debug but it never stop in my break point....
So, sorry to digress. Must we add classifier in QuarkusExtDependency by adding DefaultDependencyArtifact in the AppModelGradleResolver ? I don't really know internal gradle class

aloubyansky commented:
Can you guys still reproduce the error? I am back from pto and can't seem to reproduce it.

aloubyansky commented:
i just rebased and pushed.

Dufgui commented:
Hello, yes I reproduce it. 
I simply do a `maven clean install` on your branch of quarkus.

Then I use the gatting started project from quarkus quickstart
I just do on command line in the project:
`gradle test`

aloubyansky commented:
Strange, still can't reproduce it. Started with an empty repo (also removed my `.gradle` dir completely). Built Quarkus snapshot. Getting started tests passed.
Then I removed everything from the maven repo which is not io.quarkus. Also removed my `.gradle` dir. It still worked.
This is the only diff I have
````
diff --git a/getting-started/build.gradle b/getting-started/build.gradle
index e292efc..159d133 100644
--- a/getting-started/build.gradle
+++ b/getting-started/build.gradle
@@ -27,3 +27,11 @@ dependencies {
     testImplementation 'io.quarkus:quarkus-junit5'
     testImplementation 'io.rest-assured:rest-assured'
 }
+
+test {
+    dependsOn 'cleanTest'
+    useJUnitPlatform()
+
+    // @SubstrateTest and JVM mode tests can't be mixed in the same run
+    forkEvery 1
+}
````

aloubyansky commented:
Weird, I still can't reproduce it. I see it being downloaded to `.gradle/caches/modules-2/files-2.1/com.google.inject/guice`. So I remove it every time. My local Maven repo does not include it.

>  Must we add classifier in QuarkusExtDependency by adding DefaultDependencyArtifact in the AppModelGradleResolver ?
This code is not related to the issue. This artifact is a test dependency. It should be downloaded by Gradle itself before the test is executed. AFAIU.

Dufgui commented:
I will take a look this weekend

Le jeu. 8 août 2019 à 16:40, Alexey Loubyansky <notifications@github.com> a
écrit :

> Weird, I still can't reproduce it. I see it being downloaded to
> .gradle/caches/modules-2/files-2.1/com.google.inject/guice. So I remove
> it every time. My local Maven repo does not include it.
>
> Must we add classifier in QuarkusExtDependency by adding
> DefaultDependencyArtifact in the AppModelGradleResolver ?
> This code is not related to the issue. This artifact is a test dependency.
> It should be downloaded by Gradle itself before the test is executed. AFAIU.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/quarkusio/quarkus/pull/3332?email_source=notifications&email_token=AABZ5G3XDZ6FCLHDBCYKMULQDQV6PA5CNFSM4IG25D3KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD332XRI#issuecomment-519547845>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABZ5G6GHE25GGGS6PBXM7TQDQV6PANCNFSM4IG25D3A>
> .
>

Dufgui commented:
ok, I remove the parent directory of guice-4.1.0-no_aop.jar (.../.m2/repository/com/google/inject/guice/4.1.0) and now gradle download the aop jar.
So it's a gradle problem.

so test compilation work
But 2 tests on 4 failed.

testGreetingEndpoint()

java.net.ConnectException: Connection refused (Connection refused)
...
at org.acme.quickstart.GreetingResourceTest.testGreetingEndpoint(GreetingResourceTest.java:28)

testHelloEndpoint()

org.junit.jupiter.api.extension.TestInstantiationException: TestInstanceFactory [io.quarkus.test.junit.QuarkusTestExtension] failed to instantiate test class [org.acme.quickstart.NativeGreetingResourceIT]: Quarkus native image start failed, original cause: java.io.IOException: Cannot run program "/Users/guillaumedufour/git/quarkus/quarkus-quickstarts/getting-started/build/getting-started-unspecified-runner": error=2, No such file or directory
...
Caused by: org.junit.platform.commons.JUnitException: Quarkus native image start failed, original cause: java.io.IOException: Cannot run program "/Users/guillaumedufour/git/quarkus/quarkus-quickstarts/getting-started/build/getting-started-unspecified-runner": error=2, No such file or directory

Dufgui commented:
I do a `buildNative` before `test` and it's still KO. So I think test must depends of buildNative no ?

The second error change to :

Caused by: java.lang.RuntimeException: Unable to start native image in 60000ms
	at io.quarkus.test.common.NativeImageLauncher.waitForQuarkus(NativeImageLauncher.java:136)
	at io.quarkus.test.common.NativeImageLauncher.start(NativeImageLauncher.java:67)
	at io.quarkus.test.junit.QuarkusTestExtension.createTestInstance(QuarkusTestExtension.java:333)

machi1990 commented:
> I do a `buildNative` before `test` and the native tests are still KO. But I think, maybe, `test` task must depends of `buildNative` task no ?
> 
> The second error change to :
> 
> ```
> Caused by: java.lang.RuntimeException: Unable to start native image in 60000ms
> 	at io.quarkus.test.common.NativeImageLauncher.waitForQuarkus(NativeImageLauncher.java:136)
> 	at io.quarkus.test.common.NativeImageLauncher.start(NativeImageLauncher.java:67)
> 	at io.quarkus.test.junit.QuarkusTestExtension.createTestInstance(QuarkusTestExtension.java:333)
> ```
> 
> When I execute image myself in command line :
> 
> ```
> ./build/getting-started-unspecified-runner
> zsh: exec format error: ./build/getting-started-unspecified-runner
> ```

Related https://github.com/quarkusio/quarkus/issues/3413?

Dufgui commented:
I don't know how to say I approve this MR. but it's the case.

Dufgui commented:
> > I do a `buildNative` before `test` and the native tests are still KO. But I think, maybe, `test` task must depends of `buildNative` task no ?
> > The second error change to :
> > ```
> > Caused by: java.lang.RuntimeException: Unable to start native image in 60000ms
> > 	at io.quarkus.test.common.NativeImageLauncher.waitForQuarkus(NativeImageLauncher.java:136)
> > 	at io.quarkus.test.common.NativeImageLauncher.start(NativeImageLauncher.java:67)
> > 	at io.quarkus.test.junit.QuarkusTestExtension.createTestInstance(QuarkusTestExtension.java:333)
> > ```
> > 
> > 
> > When I execute image myself in command line :
> > ```
> > ./build/getting-started-unspecified-runner
> > zsh: exec format error: ./build/getting-started-unspecified-runner
> > ```
> 
> Related #3413?

No it's because i don't read the gradle documentation. Shame on me ;o)

> The produced executable will be a 64 bit Linux executable, so depending on your operating system it may no longer be runnable. However, it’s not an issue as we are going to copy it to a Docker container.

geoand commented:
I'll give it another shot as well

geoand reviewed (approved):
Works after applying the suggestion of removing `4.1.0` of `guice` from `.m2`. 
I suppose it's some gradle intricacy.

geoand commented:
@aloubyansky feel free to merge if you are confident with the approach

aloubyansky commented:
Thanks for reviewing, guys. And thanks helping investigate all these issues @Dufgui. I am going to merge it.

aloubyansky commented:
BTW, I did think making test depend on buildNative but it's not good idea in general. Our tests are not well organized in this regard, i.e. unit tests and integration tests should be separated and be in different directories. We should think about doing that in our quickstarts.
